### PR TITLE
[FLINK-25194] Implement an API for duplicating artefacts

### PR DESCRIPTION
**Pull request by akalash**
## What is the purpose of the change

This  PR introduces the API for duplicating artefacts

## Brief change log

*(for example:)*
  - *Several new interfaces was created*

## Verifying this change

This change added tests and can be verified as follows:

  - *New test for CheckpointStateToolset*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (**yes** / no)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (**yes** / no)
  - If yes, how is the feature documented? (not applicable / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+import java.io.IOException;
+import java.util.List;
+
+/**
+ * An extension interface for {@link FileSystem FileSystems} that can perform cheap DFS side
+ * duplicate operation. Such an operation can improve the time required for creating cheaply
+ * independent snapshots from incremental snapshots.
+ */
+public interface DuplicatingFileSystem {
```
Comment (akalash) [inline flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java:29]: Right now, we get FileSystem from Path#getFileSystem. Do I understand correctly that we will need to rewrite this code as well?

Comment (dawidwys) [reply flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java:29]: Not sure, if I understand your concern. Do you mind to elaborate?

File: flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStreamFactory.java [hunk 2]
```diff
@@ -39,4 +40,31 @@
...
+     * CheckpointedStateScope}.
+     *
+     * <p>You should first check if you can duplicate with {@link
+     * #canFastDuplicate(StreamStateHandle, CheckpointedStateScope)}.
+     *
+     * @param stateHandles The handles to duplicate
+     * @param scope Scope determining the location to duplicate into
+     * @return The duplicated handle
+     */
+    List<StreamStateHandle> duplicate(
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStreamFactory.java:68]: I agree the name does not fit in here well. Do you have any suggestions?

[the remark this answers is not in the document]
Comment (akalash) [reply flink-runtime/src/main/java/org/apache/flink/runtime/state/CheckpointStreamFactory.java:68]: As I understand, this factory is a semi-public interface(it has an implementation in several places). So perhaps, it doesn't make sense to change the name right now. Considering that I don't have the better name in my mind.

File: flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java (hunk 1 above)
Comment (akalash) [reply flink-core/src/main/java/org/apache/flink/core/fs/DuplicatingFileSystem.java:29]: My question was how does this interface will be used. I thought that we are using FileSystem implementation from a third-party library but now I understand we have our own implementation as well so it should not be a problem to expand the such implementation with this interface.

**Review by dawidwys — APPROVED** [review]
LGTM

**Pull request merged**